### PR TITLE
Sharing support

### PR DESCRIPTION
**Pull request by jon8787**

### CODE REVIEW

File: test/UID2.Client.Test/EncryptionTestsV3.cs [hunk 1]
```diff
@@ -32,6 +32,154 @@ public void SmokeTest()
             Assert.Equal(EXAMPLE_UID, res.Uid);
         }
 
+        //////////////////////  Sharing tests
+        [Fact]
+        public void CanEncryptAndDecryptForSharing()
```
Comment (sunnywu) [inline test/UID2.Client.Test/EncryptionTestsV3.cs]: this test verifies if the same client instance can encrypt and decrypt. shall we have other tests with 2 clients and 1 client encryption's can be decrypted by another client instance's decryption call? (or that part of sharing is out of the scope of this MR)?

Comment (jon8787) [reply test/UID2.Client.Test/EncryptionTestsV3.cs]: done, thanks

File: test/UID2.Client.Test/EncryptionTestsV3.cs [hunk 2]
```diff
@@ -32,6 +32,154 @@ public void SmokeTest()
...
+            Assert.Equal(EncryptionStatus.Success, encrypted.Status);
+            string advertisingToken = encrypted.EncryptedData;
+
+            var res = client.Decrypt(advertisingToken, NOW);
+            Assert.True(res.Success);
+            Assert.Equal(EXAMPLE_UID, res.Uid);
+        }
+        /*
+        [Fact]
+        public void EncryptKeyExpired()
```
Comment (sunnywu) [inline test/UID2.Client.Test/EncryptionTestsV3.cs]: just confirming we wanna keep these commented after merge?

Comment (jon8787) [reply test/UID2.Client.Test/EncryptionTestsV3.cs]: oops, fixed

File: src/UID2.Client/UID2Encryption.cs [hunk 3]
```diff
@@ -168,7 +167,45 @@ private static DecryptionResponse DecryptV3(byte[] encryptedId, IKeyContainer ke
...
+                return EncryptionDataResponse.MakeError(EncryptionStatus.NotAuthorizedForKey);
+            }
+
+            var expiry = now.AddSeconds(keys.TokenExpirySeconds);
+            var encryptParams = UID2TokenGenerator.DefaultParams.WithTokenExpiry(expiry);
+
+            try
+            {
+                string advertisingToken = (identityScope == IdentityScope.UID2) ? UID2TokenGenerator.GenerateUid2TokenV4(rawUid, masterKey, keys.CallerSiteId, defaultKey, encryptParams) :
+                    UID2TokenGenerator.GenerateEuidTokenV4(rawUid, masterKey, keys.CallerSiteId, defaultKey, encryptParams);
```
Comment (sunnywu) [inline src/UID2.Client/UID2Encryption.cs:198]: actually if we set the identityScope into encryptParams directly, we don't even need GenerateEuidTokenV4 method?

File: src/UID2.Client/UID2Encryption.cs [hunk 4]
```diff
@@ -168,7 +167,45 @@ private static DecryptionResponse DecryptV3(byte[] encryptedId, IKeyContainer ke
             }
         }
 
-        internal static EncryptionDataResponse EncryptData(EncryptionDataRequest request, IKeyContainer keys, IdentityScope identityScope)
+        internal static EncryptionDataResponse Encrypt(string rawUid, KeyContainer keys, IdentityScope identityScope, DateTime now)
```
Comment (sunnywu) [inline src/UID2.Client/UID2Encryption.cs:170]: should we make IdentityType a param and add it as a param for the public-facing UID2.Client.UID2Client.Encrypt method? currently it's not set and just default to email in the Param constructor

File: src/UID2.Client/UID2Encryption.cs (hunk 3 above)
Comment (jon8787) [reply src/UID2.Client/UID2Encryption.cs:198]: true - but doesn't matter too much as this is not public-facing

File: src/UID2.Client/IUID2Client.cs [hunk 5]
```diff
@@ -34,10 +35,15 @@ public interface IUID2Client
         /// <returns>Response showing if decryption is successful and the resulting UID if successful.
         /// Or it could return error codes/string indicating what went wrong
         /// </returns>
+        [Obsolete("Please use Decrypt(string token) instead.")] 
         DecryptionResponse Decrypt(string token, DateTime now);
```
Comment (sunnywu) [inline src/UID2.Client/IUID2Client.cs]: should we rename the param to utcNow to make it obvious that it should be utc and state that in the param comment? and same for encrypt method

File: src/UID2.Client/UID2Encryption.cs (hunk 4 above)
Comment (jon8787) [reply src/UID2.Client/UID2Encryption.cs:170]: fixed by making IdentityType derived from rawUid

File: src/UID2.Client/IUID2Client.cs (hunk 5 above)
Comment (jon8787) [reply src/UID2.Client/IUID2Client.cs]: can change the param but this is Obsolete, and Encrypt() in this class doesn't take a DateTime

Comment (sunnywu) [reply src/UID2.Client/IUID2Client.cs]: ah ok, i suggest to change the param name to make it clear for the 2 people out there but anyway nitpick

**Pull request merged**